src/librustc_trans/mir/place.rs

```diff
@@ -91,27 +91,16 @@ impl<'a, 'tcx> PlaceRef<'tcx> {
         }
 
         let scalar_load_metadata = |load, scalar: &layout::Scalar| {
-            let (min, max) = (scalar.valid_range.start, scalar.valid_range.end);
-            let max_next = max.wrapping_add(1);
-            let bits = scalar.value.size(bx.cx).bits();
-            assert!(bits <= 128);
-            let mask = !0u128 >> (128 - bits);
-            // For a (max) value of -1, max will be `-1 as usize`, which overflows.
-            // However, that is fine here (it would still represent the full range),
-            // i.e., if the range is everything.  The lo==hi case would be
-            // rejected by the LLVM verifier (it would mean either an
-            // empty set, which is impossible, or the entire range of the
-            // type, which is pointless).
-            match scalar.value {
-                layout::Int(..) if max_next & mask != min & mask => {
-                    // llvm::ConstantRange can deal with ranges that wrap around,
-                    // so an overflow on (max + 1) is fine.
-                    bx.range_metadata(load, min..max_next);
-                }
-                layout::Pointer if 0 < min && min < max => {
-                    bx.nonnull_metadata(load);
+            if let Some(range) = scalar.range_metadata(bx.cx) {
+                match scalar.value {
+                    layout::Int(..) => {
+                        bx.range_metadata(load, range);
+                    }
+                    layout::Pointer if 0 < range.start && range.start < range.end => {
```

Comment:
Not sure whether this matters, but this comparison changed. `range.end` is equal to the old `max_next`, not `max`

Reply:
How many things can I inadvertently change in a single piece of code? I think this actually disables a bunch of `nonnull` metadata by mistake, but it shouldn't be causing that type checking failure. Will fix anyway.